src/model/mod.rs

```diff
@@ -20,6 +21,17 @@ glib::wrapper! {
     pub struct DeviceModel(ObjectSubclass<imp::DeviceModel>) @implements gio::ListModel;
 }
 
+impl TrafficModel {
+    pub fn update(&self) {
+        let mut tree_opt  = self.imp().tree.borrow_mut();
+        let tree = tree_opt.as_mut().unwrap();
+        if let Some((position, _, added)) = tree.update().unwrap() {
```

Comment:
The first `unwrap` here is fine, since that `Option` is just there for annoying GObject reasons and we know it can't actually be `None`.

I think the second one should still be moved one step further out, into the idle handle closure in `main.rs`.